File: python/Arrays/day27/maximumProdSub.py
n = int(input("Enter number of element: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))



# -------- BRUTE FORCE APPROACH ---------
# maxi = float('-inf')

# for i in range(n):
#     for j in range(i, n):
#         prod = 1
#         for k in range(i, j + 1):
#             prod = prod * arr[k]
            
#         maxi = max(maxi, prod)





# -------- BETTER APPROACH ---------
# maxi = float('-inf')

# for i in range(n):
#     prod = 1
#     for j in range(i, n):
#         prod = prod * arr[j]
#         maxi = max(maxi, prod)





# -------- OPTIMAL APPROACH ---------
maxi = float('-inf')
prefix, suffix = 1, 1

for i in range(n):
    if prefix == 0:
        prefix = 1
    if suffix == 0:
        suffix = 1
    
    prefix *= arr[i]
    suffix *= arr[n - i - 1]

    maxi = max(maxi, max(prefix, suffix))





print("Maximum Product Subarray =", maxi)

File: java/Arrays/Day27/MaximumProdSub.java
package Day27;
import java.util.*;

public class MaximumProdSub {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        System.out.print("Enter " + n + " " + "numbers: ");
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }



        // -------- BRUTE APPROACH FOR MAXIMUM PRODUCT SUBARRAY ----------
        // int maxi = Integer.MIN_VALUE;
        
        // for (int i = 0; i < n; i++) {
        //     for (int j = i; j < n; j++) {
        //         int prod = 1;
        //         for (int k = i; k <= j; k++) {
        //             prod = prod * arr[k];
        //         }

        //         maxi = Math.max(maxi, prod);
        //     }
        // }





        // -------- BETTER APPROACH ----------
        // int maxi = Integer.MIN_VALUE;
        
        // for (int i = 0; i < n; i++) {
        //     int prod = 1;
        //     for (int j = i; j < n; j++) {
        //         prod *= arr[j];
        //         maxi = Math.max(maxi, prod);
        //     }
        // }
        







        // -------- OPTIMAL APPROACH ----------
        int maxi = Integer.MIN_VALUE;
        int prefix = 1, suffix = 1;
        for (int i = 0; i < n; i++) {
            if (prefix == 0) prefix = 1;
            if (suffix == 0) suffix = 1;

            prefix = prefix * arr[i];
            suffix = suffix * arr[n - i - 1];

            maxi = Math.max(maxi, Math.max(prefix, suffix));
        }





        System.out.print("Maximum product of subarray: " + maxi);

        sc.close();
    }
}
